pkg/querier/vcs/client/metrics.go

```diff
@@ -0,0 +1,103 @@
+package client
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/go-github/v58/github"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// Metrics can record events surrounding various client actions, like logging in
+// or fetching a file.
+type Metrics interface {
+	LoginObserve(elapsed time.Duration, err error)
+	RefreshObserve(elapsed time.Duration, err error)
+	GetCommitObserve(elapsed time.Duration, res *github.Response, err error)
+	GetFileObserve(elapsed time.Duration, res *github.Response, err error)
+}
+
+func NewMetrics(reg prometheus.Registerer) Metrics {
+	return &githubMetrics{
+		APIDuration: promauto.With(reg).NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "pyroscope",
+				Name:      "vcs_github_request_duration",
+				Help:      "Duration of GitHub API requests in seconds",
+				Buckets:   prometheus.ExponentialBucketsRange(0.1, 10, 8),
+			},
+			[]string{"path", "status"},
+		),
+		APIRateLimit: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
```

Comment:
I worry that this metric might not be too helpful/accurate:

My understanding is that we are subject to rate limits of the github user that authenticated via oauth2. So if we have two github users using the service at the same time we would constantly have a jumping value, between those two values of remaining requests.

I do think this is something important to watch. Maybe this could be a histogram with fitting buckets, that would show us how close to being rate limited we got.

Reply:
You're absolutely right here. The rate limit remaining here is _per-user_ ([doc](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-authenticated-users)) since we're using a user token. If we had an installation token, the rate limit would be per app.

How I'm using the metric here is not helpful at all. Like you said, it's going to jump to whatever remaining rate limit the last token had. I'm going to rethink how to track this, as a gauge is clearly not the right approach.